File: lecture_notes/09 functions/a-test.js
'use strict';

const cache = {};

const addProcedure = (a, b) => {
  const key = `${a},${b}`;
  let res = cache[key];
  if (res) return res;
  res = a + b;
  cache[key] = res;
  return res;
};

const subProcedure = (a, b) => {
  const key = `${a},${b}`;
  let res = cache[key];
  if (res) return res;
  res = a - b;
  cache[key] = res;
  return res;
};

const test = (f, args, res) => {
  const result = f(...args);
  if (result === res) return true;
  else return false;
};

console.log([
  test(addProcedure, [1, 2], 3),
  test(addProcedure, [100, 20], 120),
  test(addProcedure, [100, 200], 300),
  test(subProcedure, [5, 2], 3),
  test(addProcedure, [5, 2], 7),
]);

File: lecture_notes/09 functions/a-test.py
cache = {}


def add_procedure(a, b):
    key = f'{a}, {b}'
    if key in cache:
        return cache[key]
    res = a + b
    cache[key] = res
    return res


def sub_procedure(a, b):
    key = f'{a}, {b}'
    if key in cache:
        return cache[key]
    res = a - b
    cache[key] = res
    return res


def test(f, args, res):
    result = f(*args)
    if result == res:
        return True
    else:
        return False


print([
    test(add_procedure, [1, 2], 3),
    test(add_procedure, [100, 20], 120),
    test(add_procedure, [100, 200], 300),
    test(sub_procedure, [5, 2], 3),
    test(add_procedure, [5, 2], 7)
])
